### ISSUE TITLE
False positive with no-implied-eval

### ISSUE THREAD

**Issue by davidgilbertson** [typescript-eslint/typescript-eslint#1642]
**What code were you trying to parse?**
Copy pasted from the readme: https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-implied-eval.md
```ts
const fn = () => {};
setTimeout(fn, 100);
```

**What did you expect to happen?**
No error

**What actually happened?**
```
ESLint: Implied eval. Consider passing a function.(@typescript-eslint/no-implied-eval)
```

**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `2.21.0` |
| `TypeScript`                | `3.8.2` |
| `ESLint`                    | `6.8.0` |
| `node`                      | `12.14.0` |
| `npm`                       | `6.13.4` |

**Comment by bradzacher** [comment]
I'm unable to repro this against master.
It's working fine in both tests and inline.

Could you please provide some more information about your environment?

**Comment by g1thuser** [comment]
I can confirm this issue. The package versions for @typescript-eslint/parser, TypeScript and ESLint are the same here. My config files are highly customized and it is difficult to track it down, which setting causes the problem, or to provide it properly here, but the probably relevant sections are looking like this:

eslintrc.js:

```
module.exports = {
  parserOptions: {
      "ecmaVersion": 8,
      "project": tsconfig.json,
  },
  plugins: [
    "dollar-sign",
    "eslint-comments",
    "extra-rules",
    "jsdoc",
    "more",
    "optimize-regex",
    "prettiest",
    "sonarjs",
    "switch-case",
    "unicorn",
    "@getify/proper-arrows",
    "@typescript-eslint",
    "array-func",
    "import",
    "import-helpers",
    "no-constructor-bind",
    "prefer-arrow",
    "promise",
    "sort-class-members",
    "sort-destructure-keys",
  ],
  overrides: [
    {
      files: ["*.ts"],
      parser: "@typescript-eslint/parser",
      parserOptions: {
        "sourceType": "module",
      },
  ],
  env: {
    "es2017": true,
    "node": true,
  },
}
```
tsconfig.json:

```
"compilerOptions": {
  "lib": [
      "dom",
      "es5",
      "es2015",
      "es2015.iterable",
      "es2015.promise",
      "es2015.symbol",
      "es2015.symbol.wellknown"
  ],
  "module": "system",
  "moduleResolution": "node",
  "target": "es2015",
}
```

**Comment by bradzacher** [comment]
If you could distill it down to a simple repro, and chuck that in a github gist or repo, then I would be better able to look into this.
With it passing fine in our tests, I'm not sure what the problem could be.

**Comment by yenshih** [comment]
Exact same issue here. I'm using `plugin:@typescript-eslint/eslint-recommended` and `plugin:@typescript-eslint/all`.

**Comment by bradzacher** [comment]
Please avoid commenting with "me too", use the +1 button on op and subscribe yourself.

https://github.com/typescript-eslint/typescript-eslint/blob/master/CONTRIBUTING.md#commenting

Please only comment if you have some valuable information to give, like a repro repo.

**Comment by yeonjuan** [comment]
I checked that it's not reproduced in typescript `3.7.x`. It only occurs when a user using typescript `3.8.x`.  - [repro repo- using typescript 3.8.2](https://github.com/yeonjuan/tseslint-no-implied-eval)

Maybe it's because the typescript-eslint doesn't support typescript >= 3.8.0 for now.
I hope this information is helpful to someone. :)

**Comment by bradzacher** [comment]
Awesome, thanks for figuring that out. With that I was able to find the issue,
It seems like 3.8 has a slight change in how the type is represented, which causes the rule to falsely report arrow functions only.

**Comment by davidgilbertson** [comment]
@bradzacher thanks so much for resolving this so quickly. @yeonjuan thanks for the troubleshooting that I totally would have gotten around to at some point :)
